source/RoslynAnalyzers/AsyncAnalyzer.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Octopus.RoslynAnalyzers;
+
+using static Descriptors;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class AsyncAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+        VoidMethodsMustNotBeAsync,
+        MethodsReturningTaskMustBeAsync);
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(CheckNode, SyntaxKind.MethodDeclaration);
+    }
+
+    void CheckNode(SyntaxNodeAnalysisContext context)
+    {
+        if (context.Node is not MethodDeclarationSyntax methodDec) return;
+
+        var isAsync = methodDec.Modifiers.Any(SyntaxKind.AsyncKeyword);
+        if (isAsync)
+        {
+            VoidMethods_MustNotBeAsync(context, methodDec);
+        }
+
+        if (methodDec.ReturnType is SimpleNameSyntax { Identifier.Text: "Task" or "ValueTask" })
+        {
+            MethodsReturningTask_MustBeAsync(context, methodDec, isAsync);
+        }
+    }
+
+    void VoidMethods_MustNotBeAsync(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec)
+    {
+        if (methodDec.ReturnType is PredefinedTypeSyntax p && p.Keyword.IsKind(SyntaxKind.VoidKeyword))
+        {
+            context.ReportDiagnostic(Diagnostic.Create(VoidMethodsMustNotBeAsync, methodDec.Identifier.GetLocation()));
+        }
+    }
+
+    void MethodsReturningTask_MustBeAsync(SyntaxNodeAnalysisContext context, MethodDeclarationSyntax methodDec, bool isAsync)
+    {
+        // don't flag things that are async (that's good!) or abstract methods.
+        if (isAsync || methodDec.Modifiers.Any(SyntaxKind.AbstractKeyword)) return;
+
+        var declaringType = methodDec.Parent;
+        while (declaringType is not null && declaringType is not TypeDeclarationSyntax) declaringType = declaringType.Parent;
```

Comment:
This confused me for a bit - recommend expanding it out
```suggestion
        while (declaringType is not null && declaringType is not TypeDeclarationSyntax) 
        {
            declaringType = declaringType.Parent;
        }
```

Reply:
Good point. I think I'll actually take this out into an extension method like we have with GetNamespace()